FYI: Error message changed in redis between 5.0.3 and 5.0.7

stream_test.go:10: error error. expected: "ERR The ID specified in XADD must be greater than 0-0" got: "ERR The ID specified in XADD is equal or smaller than the target stream top item" case: main.command{cmd:"XADD", args:[]interface {}{"newplanets", "0", "foo", "bar"}, error:true, sort:false, loosely:false, errorSub:"", receiveOnly:false, roundFloats:0, closeChan:false, noResultCheck:false}

On Sun, Feb 16, 2020 at 02:06:11PM -0800, cleroux wrote:
> stream_test.go:10: error error. expected: "ERR The ID specified in XADD must be greater than 0-0" got: "ERR The ID specified in XADD is equal or smaller than the target stream top item" case: main.command{cmd:"XADD", args:[]interface {}{"newplanets", "0", "foo", "bar"}, error:true, sort:false, loosely:false, errorSub:"", receiveOnly:false, roundFloats:0, closeChan:false, noResultCheck:false}

Thanks! I'll update it in a few days.

--- a/README.md
+++ b/README.md
@@ -299,7 +299,7 @@ Commands which will probably not be implemented:
 
 ## &c.
 
-Tests are run against Redis 5.0.3. The [./integration](./integration/) subdir
+Tests are run against Redis 5.0.7. The [./integration](./integration/) subdir
 compares miniredis against a real redis instance.
 
 If you want to test Redis Sentinel have a look at [minisentinel](https://github.com/Bose/minisentinel).

--- a/cmd_stream.go
+++ b/cmd_stream.go
@@ -90,6 +90,8 @@ func (m *Miniredis) cmdXadd(c *server.Peer, cmd string, args []string) {
 			switch err {
 			case errInvalidEntryID:
 				c.WriteError(msgInvalidStreamID)
+			case errZeroStreamValue:
+				c.WriteError(msgStreamIDZero)
 			case errInvalidStreamValue:
 				c.WriteError(msgStreamIDTooSmall)
 			default:

--- a/cmd_stream_test.go
+++ b/cmd_stream_test.go
@@ -43,7 +43,7 @@ func TestStream(t *testing.T) {
 
 	t.Run("direct usage", func(t *testing.T) {
 		_, err := s.XAdd("s1", "0-0", []string{"name", "foo"})
-		mustFail(t, err, errInvalidStreamValue.Error())
+		mustFail(t, err, errZeroStreamValue.Error())
 
 		id, err := s.XAdd("s1", "12345-67", []string{"name", "bar"})
 		ok(t, err)

--- a/db.go
+++ b/db.go
@@ -589,7 +589,10 @@ func (db *RedisDB) streamAdd(key, entryID string, values []string) (string, erro
 	if err != nil {
 		return "", err
 	}
-	if entryID == "0-0" || streamCmp(stream.lastID(), entryID) != -1 {
+	if entryID == "0-0" {
+		return "", errZeroStreamValue
+	}
+	if streamCmp(stream.lastID(), entryID) != -1 {
 		return "", errInvalidStreamValue
 	}
 	db.streamKeys[key] = append(stream, StreamEntry{

--- a/integration/get_redis.sh
+++ b/integration/get_redis.sh
@@ -2,7 +2,7 @@
 
 set -eu
 
-VERSION=5.0.5
+VERSION=5.0.7
 
 rm -rf ./redis_src/
 mkdir -p ./redis_src/

--- a/redis.go
+++ b/redis.go
@@ -34,6 +34,7 @@ const (
 	msgSingleElementPair  = "ERR INCR option supports a single increment-element pair"
 	msgInvalidStreamID    = "ERR Invalid stream ID specified as stream command argument"
 	msgStreamIDTooSmall   = "ERR The ID specified in XADD is equal or smaller than the target stream top item"
+	msgStreamIDZero       = "ERR The ID specified in XADD must be greater than 0-0"
 	msgNoScriptFound      = "NOSCRIPT No matching script. Please use EVAL."
 	msgUnsupportedUnit    = "ERR unsupported unit provided. please use m, km, ft, mi"
 	msgNotFromScripts     = "This Redis command is not allowed from scripts"

--- a/stream.go
+++ b/stream.go
@@ -13,6 +13,7 @@ import (
 
 var (
 	errInvalidStreamValue = errors.New("stream id is not bigger than the top item")
+	errZeroStreamValue    = errors.New("stream id is 0-0")
 )
 
 type streamKey []StreamEntry

